packages/cloudflare/src/kv.ts

```diff
@@ -0,0 +1,104 @@
+import type { StorageAdapter } from 'grammy/web';
+import type { KVNamespace, KVNamespaceListResult, KVNamespaceListOptions } from '@cloudflare/workers-types';
+
+export class KvAdapter<T> implements StorageAdapter<T> {
+  private kv_namespace: KVNamespace<string>;
+
+  constructor(kv_namespace: KVNamespace<string>) {
+    this.kv_namespace = kv_namespace;
+  }
+
+  async read(key: string): Promise<T | undefined> {
+    return await this.kv_namespace.get<T>(key, { type: 'json' }) ?? undefined;
+  }
+
+  async write(key: string, value: T): Promise<void> {
+    await this.kv_namespace.put(key, JSON.stringify(value));
+  }
+
+  async delete(key: string): Promise<void> {
+    await this.kv_namespace.delete(key);
+  }
+
+  async has(key: string): Promise<boolean> {
+    let keyList: KVNamespaceListResult<any, string>;
+    const listOptions: KVNamespaceListOptions = {};
+    do {
+      keyList = await this.kv_namespace.list(listOptions);
+      if (!keyList.list_complete) {
+        listOptions.cursor = keyList.cursor;
+      }
+      for (const k of keyList.keys) {
+        if (key === k.name) return true;
+      }
+    } while (!keyList.list_complete);
+    return false;
+  }
+
+  readAllKeys(): AsyncIterable<string> {
+    const kv_namespace = this.kv_namespace;
+    return {
+      async *[Symbol.asyncIterator](): AsyncIterator<string, any, undefined> {
+        let keyList: KVNamespaceListResult<any, string>;
+        const listOptions: KVNamespaceListOptions = {};
+        do {
+          keyList = await kv_namespace.list(listOptions);
+          if (!keyList.list_complete) {
+            listOptions.cursor = keyList.cursor;
+          }
+          for (const key of keyList.keys) {
+            yield key.name;
+          }
+        } while (!keyList.list_complete);
+      },
+    };
+  }
+
+  readAllValues(): AsyncIterable<T> {
+    const kv_namespace = this.kv_namespace;
+    return {
+      async *[Symbol.asyncIterator](): AsyncIterator<T, any, undefined> {
```

Comment:
Instead of doing this, we could just make the entire function a generator. That would look much cleaner.

Reply:
Ha-ha, i'm not familar with generators, sorry.

Reply:
Okay, I fixed it up for you.

Reply:
> Okay, I fixed it up for you.

Thanks.